executor/compact_table.go

```diff
@@ -0,0 +1,337 @@
+// Returns when compaction is finished. When there are network problems it will retry internally.
+//
+// There are two kind of errors may be returned:
+// A. Error only cancel tasks related with this store, e.g. this store is down even after retry.
+// 		The remaining partitions in this store should be cancelled.
+// B. Error that should cancel tasks of other stores, e.g. CompactErrorCompactInProgress.
+//		The remaining partitions in this store should be cancelled, and tasks of other stores should also be cancelled.
+//
+// During this function, some "problems" will cause it to early return, e.g. physical table not exist in this
+// store any more (maybe caused by DDL). No errors will be produced so that remaining partitions will continue
+// being compacted.
+func (task *storeCompactTask) compactOnePhysicalTable(physicalTableID int64) (stopAllTasks_ bool, err_ error) {
+	var startKey []byte = nil
+	for { // This loop is to compact incrementally for all data. Each RPC request will only compact a partial of data.
+		if task.ctx.Err() != nil {
+			return true, task.ctx.Err()
+		}
+
+		task.logProgressOptionally()
+
+		resp, err := task.sendRequestWithRetry(&tikvrpc.Request{
+			Type:    tikvrpc.CmdCompact,
+			StoreTp: tikvrpc.TiFlash,
+			Req: &kvrpcpb.CompactRequest{
+				LogicalTableId:  task.parentExec.tableInfo.ID,
+				PhysicalTableId: physicalTableID,
+				StartKey:        startKey,
+			},
+		})
+		if err != nil {
+			// Even after backoff, the request is still failed.., or the request is cancelled or timed out
+			warn := errors.Errorf("compact on store %s failed: %v", task.targetStore.Address, err)
+			task.parentExec.ctx.GetSessionVars().StmtCtx.AppendWarning(warn)
+			task.logFailure(
+				zap.Int64("physical-table-id", physicalTableID),
+				zap.Error(err))
+			return false, warn
+		}
+		if resp.GetError() != nil {
+			switch resp.GetError().GetError().(type) {
+			case *kvrpcpb.CompactError_ErrCompactInProgress:
+				warn := errors.Errorf("compact on store %s failed: table is compacting in progress", task.targetStore.Address)
+				task.parentExec.ctx.GetSessionVars().StmtCtx.AppendWarning(warn)
+				task.logFailure(
+					zap.Int64("physical-table-id", physicalTableID),
+					zap.Error(warn))
+				// TiFlash reported that there are existing compacting for the same table.
+				// We should stop the whole SQL execution, including compacting requests to other stores, as repeatedly
+				// compacting the same table is a waste of resource.
+				return true, warn
+			case *kvrpcpb.CompactError_ErrTooManyPendingTasks:
+				// The store is already very busy, don't retry.
+				warn := errors.Errorf("compact on store %s failed: store is too busy", task.targetStore.Address)
+				task.parentExec.ctx.GetSessionVars().StmtCtx.AppendWarning(warn)
+				task.logFailure(
+					zap.Int64("physical-table-id", physicalTableID),
+					zap.Error(warn))
+				return false, warn
```

Comment:
would this tiflash replica remain uncompacted if it just gives up?

Reply:
Yes! We will let user know that store is too busy and table is not compacted (via warnings). User can decide to compact later again.